ui/components/app/snaps/snap-ui-renderer/snap-ui-renderer.js

```diff
@@ -53,11 +89,86 @@ const SnapUIRendererComponent = ({
     // We do this to avoid useless re-renders.
     (oldState, newState) => isEqual(oldState.content, newState.content),
   );
-  const rawContent = interfaceState?.content;
-  const content =
-    rawContent?.type === 'Container' || !rawContent
+
+  const content = useMemo(() => {
+    const rawContent = interfaceState?.content;
+    return rawContent?.type === 'Container' || !rawContent
       ? rawContent
       : Container({ children: rawContent });
+  }, [interfaceState?.content]);
+
+  const requireScroll = useMemo(
+    () => Boolean(content?.props?.children?.[1]?.props?.requireScroll),
+    [content?.props?.children],
+  );
+
+  const isProgrammaticScrollRef = useRef(false);
+
+  const [scrollState, dispatch] = useReducer(scrollReducer, {
+    isProgrammaticScroll: false,
+    buttonsEnabled: !requireScroll,
+    showArrow: false,
+    hasReachedBottom: !requireScroll,
+  });
+
+  const { onScroll, scrollToBottom, isScrollable, ref } = useScrollRequired(
+    [],
+    {
+      // Only enable the hook if we need to handle scroll events
+      enabled: requireScroll,
+      // This lets us batch state updates and avoid an unnecessary render
+      onMeasure: ({ isScrollable: canScroll, hasMeasured }) => {
+        if (!requireScroll || !hasMeasured) {
+          return;
+        }
+        dispatch({
+          type: 'MANUAL_SCROLL',
+          isScrollable: canScroll,
+          isAtBottom: !canScroll,
+        });
+      },
+    },
+  );
+
+  const buttonsEnabled = useMemo(
+    () => (requireScroll ? scrollState.buttonsEnabled : true),
+    [requireScroll, scrollState.buttonsEnabled],
+  );
+
+  const handleScroll = useCallback(
+    (e) => {
+      if (!requireScroll) {
+        return;
+      }
+
+      const isActuallyAtBottom =
+        Math.abs(
+          e.target.scrollTop + e.target.clientHeight - e.target.scrollHeight,
+        ) < 2;
+
+      if (isProgrammaticScrollRef.current && isActuallyAtBottom) {
+        isProgrammaticScrollRef.current = false;
+        dispatch({ type: 'COMPLETE_SCROLL' });
+      } else {
+        dispatch({
+          type: 'MANUAL_SCROLL',
+          isAtBottom: isActuallyAtBottom,
+          isScrollable,
+        });
+      }
+      onScroll?.(e);
+    },
+    [onScroll, isScrollable, requireScroll],
+  );
+
+  const handleScrollToBottom = useCallback(() => {
+    // Prevent multiple clicks
+    if (isProgrammaticScrollRef.current) {
+      return;
+    }
+    isProgrammaticScrollRef.current = true;
```

Comment:
Why is this a ref? Is this necessary in the first place?

Reply:
Glad you questioned this, made me refactor things a little. Removed the reducer state and simplified the scroll handler. The ref is needed for the scroll to bottom function, I don't want to use state to set a flag and cause an unnecessary re-render.